### PR TITLE
Tensorize dynamic inference mixed sampling

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?
#1927 implemented mixed sampling in an unoptimized fashion. This led to a significant regression in the latency of the forward step.

This PR tensorizes the associated bookkeeping, eliminating the related slowdown.

A sketch of the design follows:

- SamplingParams now uses `unsafe_hash=True`. A future PR should make it immutable and use `frozen=True`.
- The dynamic context now has an extra request-indexed Tensor, which keeps track of the SamplingParams hash associated with each request in the context. Every time a request is added to the dynamic context, its SamplingParams are hashed and associated with the new request.
- When the engine steps, there is no longer a need to use a slow nested Python loop to break up requests into SamplingParams buckets. Instead, this now uses tensorized logic inside `_dynamic_step_sample_bookkeeping`.
- An optional `static_sampling` parameter can be used to bypass even this tensorized logic.
- The sampling code from `sample_from_logits` has been factored out into is own sub-method, but neither it nor any of the `sample_from_logits` code has been modified at all.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [x] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/sampling_params.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 from typing import Optional
 
 
-@dataclass
+@dataclass(unsafe_hash=True)
```
Comment (tdene) [inline megatron/core/inference/sampling_params.py]: This is one of the main things I want reviewed.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 2]
```diff
@@ -339,10 +340,15 @@ def bytes_to_max_requests_and_tokens(n_bytes):
         self.padded_active_request_count = None
         self.paused_tokens = None
 
+        # Keep a reverse hash map for SamplingParams that correspond to active & paused requests.
+        self.sampling_hash_map: Dict[int, SamplingParams] = {}
+
```
Comment (tdene) [inline megatron/core/inference/contexts/dynamic_context.py]: This is one of the main things I want reviewed.

Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py]: Specifically, is this kind of thing out of scope for the dynamic context? Do I need to redesign how this data gets tracked and updated over time?

**Comment by tdene** [comment]
All tests pass.

Merge conflict is due to branches temporarily being reverted from main, and it will soon be resolved on its own.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 3]
```diff
@@ -1197,6 +1203,19 @@ def add_request(self, req: DynamicInferenceRequest, chunk_length: Optional[int]
...
+        core_params = req.sampling_params.core_params
+        # Handle the sampling parameters hash.
+        old_sampling_hash = self.request_sampling_hashes[current_id].item()
+        new_sampling_hash = hash(core_params)
+        self.request_sampling_hashes[current_id] = new_sampling_hash
+        self.sampling_hash_map[new_sampling_hash] = core_params
+        # Update the reverse hash map if there are no more requests with this sampling hash.
+        # We do this to prevent unbounded growth of the hash map.
+        if (self.request_sampling_hashes != old_sampling_hash).all().item():
+            if old_sampling_hash in self.sampling_hash_map:
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py]: is it possible for `old_sampling_hash` to not be in `sampling_hash_map` at this point?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 4]
```diff
@@ -1197,6 +1203,19 @@ def add_request(self, req: DynamicInferenceRequest, chunk_length: Optional[int]
...
+        # Get immutable core of sampling_params.
+        core_params = req.sampling_params.core_params
+        # Handle the sampling parameters hash.
+        old_sampling_hash = self.request_sampling_hashes[current_id].item()
+        new_sampling_hash = hash(core_params)
+        self.request_sampling_hashes[current_id] = new_sampling_hash
+        self.sampling_hash_map[new_sampling_hash] = core_params
+        # Update the reverse hash map if there are no more requests with this sampling hash.
+        # We do this to prevent unbounded growth of the hash map.
+        if (self.request_sampling_hashes != old_sampling_hash).all().item():
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py]: I think it's a bit expensive to compare all request's hashes here. What if we add another map to the context of <hash, set(request_ids)> where we add the request_id in `add_request` and then remove the request_id in `update_request` if it's finished. Then, when deciding whether to remove `old_sampling_hash` here, we simply check if the `set(request_ids)` is empty.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 5]
```diff
@@ -567,31 +547,65 @@ def _dynamic_step_forward_logits(self, input_ids: Tensor, position_ids: Tensor)
...
+
+        # Special-case for static sampling.
+        if getattr(self, "static_sampling", False):
+            assert len(sampling_hash_map) == 1
+            core_params = next(iter(sampling_hash_map.values()))
+            self.torch_sampling_groups[core_params] = None
+            return
+
+        # Loop over all unique sampling hashes.
+        for sampling_hash, core_params in sampling_hash_map.items():
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: I believe it's a bit inefficient to iterate over the entire `sampling_hash_map`, rather than to only focus on active hashes. I would group together the active hashes by their hash value, a mapping like <hash, set(active_ids)>, and then loop over this. Otherwise, as it is now we'll have iterations where `indices` is empty on the next line.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 4 above)
Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py]: I like it!

File: megatron/core/inference/contexts/dynamic_context.py [hunk 6]
```diff
@@ -442,6 +446,16 @@ def bytes_to_max_requests_and_tokens(n_bytes):
         # request_last_kv_block_offset represents number of tokens in the last kv block
         self.request_last_kv_block_offset = torch.empty_like(self.request_ids)
 
+        # Track request metadata.
+        if num_request_metadata is None:
+            num_request_metadata = len(DynamicInferenceRequest.get_metadata_labels())
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:451]: I'm still a bit concerned about reading `DynamicInferenceRequest` directly ... what happens if someone is using a custom `MyOneBigBeautifulRequest` and then the wrong `get_metadata_labels()` is used here?

Would there be any issue with defaulting to `num_request_metadata = 0` or `None` here?

File: megatron/core/inference/contexts/dynamic_context.py
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:1]: I think we should also reset `self.request_metadata.fill_(0)` inside of `def reset()`

File: megatron/core/inference/inference_request.py [hunk 7]
```diff
@@ -277,6 +277,46 @@ def deserialize(cls, obj: dict) -> "DynamicInferenceRequest":
...
+            sp.top_p,
+            sp.termination_id,
+            sp.return_log_probs,
+            getattr(sp, "skip_prompt_log_probs_for_dynamic_inference", False),
+        ]
+
+    @staticmethod
+    def get_metadata_labels() -> Dict[str, int]:
+        """Provides human-readable labels for the tracked metadata fields."""
+        ret = [
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/inference_request.py:303]: can we merge this list with the `sp.*` list above by using `getattr`, just or easier maintainability

File: megatron/core/inference/engines/dynamic_engine.py [hunk 8]
```diff
@@ -655,37 +661,6 @@ def schedule_non_chunked_prefill(self):
...
-        ].tolist()
-        if self.static_sampling:
-            return [(next(iter(self.requests.values())).sampling_params, active_request_ids)]
-
-        # Get a map from request_id to context array index.
-        context_id_map = {r: i for i, r in enumerate(active_request_ids)}
-
-        # Create map of sampling methods to context array indices.
-        sampling_map: List[Tuple[SamplingParams, List[int]]] = []
-        for request_id, request in self.requests.items():
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:672]: I would recommend iterating over `active_request_ids`, to avoid the check with `context_id_map`.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 9]
```diff
@@ -655,37 +661,6 @@ def schedule_non_chunked_prefill(self):
...
-            if request_id not in context_id_map:
-                continue
-            context_id = context_id_map[request_id]
-            sp = request.sampling_params
-
-            # Look for a pre-existing group with these sampling parameters.
-            for sampling, indices in sampling_map:
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py:679]: I think this nested for-loop can/should be avoided. Before the loop, we can initialize `sampling_map = defaultdict(set)` that will hold a mapping of `<hash, set(request_ids)>`. Then we iterate just over `active_request_ids`, grab that request's sampling hash, and directly add it with `sampling_map[sampling_hash].add(active_request_id)`. This avoids a lot of extra looping and comparisons.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 10]
```diff
@@ -177,45 +210,14 @@ def detokenize_generations(
...
+        temperature: float,
+        top_k: int,
+        top_p: float,
         vocab_size: Optional[int] = None,
-        generation_started: Optional[torch.Tensor] = None,
-        top_n_logprobs_dict: Dict[int, List[Dict[str, float]]] = None,
-        logits: Optional[torch.Tensor] = None,
-        **kwargs,
-    ) -> torch.Tensor:
-        """Samples the logits to generate outputs
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py:190]: why did this docstring get removed?

File: megatron/core/inference/engines/dynamic_engine.py (hunk 9 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:679]: This code has been thrown out entirely. I think I did what you suggested here, in `text_generation_controller`'s `_dynamic_step_sample_bookkeeping`:

```
        if backend == "torch":
            # Bucketize the core sampling parameters.
            core_params = torch.stack((temp, top_k, top_p), dim=1)
            _, inv_indices, cnts = torch.unique(
                core_params, dim=0, return_inverse=True, return_counts=True
            )
            order = torch.argsort(inv_indices, stable=True)
            sampling_buckets = torch.split(order, cnts.tolist())
```

File: megatron/core/inference/engines/dynamic_engine.py (hunk 8 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:672]: This code has been thrown out entirely :).

File: megatron/core/inference/inference_request.py (hunk 7 above)
Comment (tdene) [reply megatron/core/inference/inference_request.py:303]: I was waiting on #2181 to be merged before doing this...

and done! Addressed; please resolve thread.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 10 above)
Comment (tdene) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py:190]: Addressed! Please resolve thread.

File: megatron/core/inference/contexts/dynamic_context.py
Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py:1]: Addressed! Please resolve thread.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 6 above)
Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py:451]: Address over huddle.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 11]
```diff
@@ -74,6 +78,35 @@ def __init__(
...
+            max_requests, vocab_size, dtype=logits_dtype, device=device
+        )
+        self.cu_sampled_tokens = torch.empty(max_requests, dtype=torch.int64, device=device)
+
+        self.cu_temperature = torch.empty_like(self.cu_sampled_tokens, dtype=torch.float)
+        self.cu_top_k = torch.empty_like(self.cu_sampled_tokens, dtype=torch.int32)
+        self.cu_top_p = torch.empty_like(self.cu_sampled_tokens, dtype=torch.float)
+        self.cu_termination_id = torch.empty(max_requests, dtype=torch.int64, device=device)
+        self.cu_return_log_probs = torch.empty(max_requests, dtype=torch.bool, device=device)
+        self.cu_skip_prompt_log_probs = torch.empty(max_requests, dtype=torch.bool, device=device)
```
Comment (santhnm2) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: Why do these have the `cu_` prefix? Typically this means "cumulative" but in this case the values are independent across requests right?

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 11 above)
Comment (tdene) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py]: I thought cu meant `cuda`! Fixed.
